### PR TITLE
Add VolumeStoresList to volume layer

### LINKED ISSUE #1506: Add ListStores to volume layer of port layer
Return a list of volume store names.

### PR DESCRIPTION
**Pull request by fdawg4l**
NOTE:  TO BE MERGED POST BETA.

Fixes #1506 

* Added handler for port layer rest API
* Also fix sync() issues in unit test as well as detach semantics.

### CODE REVIEW

File: lib/apiservers/portlayer/swagger.yml [hunk 1]
```diff
@@ -207,6 +207,29 @@ paths:
...
+      responses:
+        '200':
+          description: "OK"
+          schema:
+            $ref: "#/definitions/VolumeStoresListResponse"
+        '404':
+          description: "Not found"
+          schema:
+            $ref: "#/definitions/Error"
+        default:
```
Comment (hmahmood) [inline lib/apiservers/portlayer/swagger.yml]: I would prefer a status code here, and not a catch all. '500' should do the trick.

Comment (fdawg4l) [inline lib/apiservers/portlayer/swagger.yml]: Sounds good.  Will do.

File: lib/portlayer/storage/vsphere/volume.go [hunk 2]
```diff
@@ -87,6 +87,28 @@ func (v *VolumeStore) AddStore(ctx context.Context, ds *datastore.Helper, storeN
 	return u, nil
 }
 
+func (v *VolumeStore) VolumeStoresList(ctx context.Context) (map[string]url.URL, error) {
+	m := make(map[string]url.URL)
+	for u, ds := range v.ds {
+
+		// Get the ds:// URL given the datastore url ("[datastore] /path")
+		dsurl, err := datastore.ToURL(ds.RootURL)
```
Comment (matthewavery) [inline lib/portlayer/storage/vsphere/volume.go]: In order to avoid copies of "ds" you could replace `ds.RootURL` with v[u].RootURL

**Comment by matthewavery** [comment]
lgtm, only one small comment. Not really a big issue.

File: lib/portlayer/storage/vsphere/volume.go (hunk 2 above)
Comment (fdawg4l) [inline lib/portlayer/storage/vsphere/volume.go]: @matthewavery are you concerned about a pointer?  ds is a map  of the following type

```
39 »···// maps datastore uri (volume store) to datastore
40 »···ds map[url.URL]*datastore.Helper 
```Copying of a pointer is trivial unless youre concerned about something else.  

Also, about your proposed solution;  in general, avoid iterating maps whenever possible. But if that can't be avoided (like here), don't then use the result of the iteration to once again look the iterated key up in the map.

**Comment by fdawg4l** [comment]
.

**Comment by caglar10ur** [comment]
LGTM :sheep: :it:

**Pull request merged**